### DISCUSSION TITLE
[6주차] 네트워크 계층: 제어 평면 5.1~4

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Dijkstra-Employ/Network-Study#14] [Q&A]
6주차 네트워크 계층: 제어 평면 5.1~4에 해당하는 질의응답입니다.

**Comment by H-sooyeon** [comment]
Q1) 네트워크 계층 라우터의 포워딩 테이블은 라우팅 알고리즘을 통해 채워집니다. 이 라우팅 알고리즘은 크게 중앙 집중형 라우팅 알고리즘과 분산 라우팅 알고리즘으로 나뉩니다. 이에 대해 각각 간단하게 설명해주세요.
<hr />

Q2) 링크 상태 알고리즘과 거리 벡터 라우팅 알고리즘의 차이에 대해 설명해주세요.
<hr />

**Reply by Starlight258** [reply]
> Q1) 네트워크 계층 라우터의 포워딩 테이블은 라우팅 알고리즘을 통해 채워집니다. 이 라우팅 알고리즘은 크게 중앙 집중형 라우팅 알고리즘과 분산 라우팅 알고리즘으로 나뉩니다. 이에 대해 각각 간단하게 설명해주세요.
> 
- 중앙 집중형 라우팅 알고리즘 : **네트워크 전체에 대한 완전한 정보**를 가지고 출발지와 목적지 사이의 최소 비용 경로를 계산한다.
  - `링크 상태 알고리즘` : 연결과 링크 비용에 대한 **완전한 정보**를 가진다.
- 분산형 라우팅 알고리즘 : 최소 비용 경로의 계산이 라우터들에 의해 **반복적이고 분산된 방식으로 수행**된다. 
  - `거리 벡터 알고리즘` :  각 노드는 자신과 직접 연결된 링크에 대한 비용 정보만 가진다.
    - `반복된 계산`과 `이웃 노드간의 정보 교환`을 통해 노드는 목적지까지의 **최소 비용 경로를 계산**한다. 

> Q2) 링크 상태 알고리즘과 거리 벡터 라우팅 알고리즘의 차이에 대해 설명해주세요.
- 링크 상태 알고리즘 : 전체 네트워크 토폴로지 정보를 모든 라우터와 공유한다.
  -  `다익스트라 알고리즘`을 사용해 최단 경로를 계산한다.
> 각 노드가 자신과 연결된 링크의 식별자와 비용을 포함하는 `링크 상태 패킷`을 네트워크 상의 모든 다른 노드로 **브로드캐스팅**한다. 이를 통해 모든 노드는 네트워크에 대한 `동일한 관점`을 갖게 된다.

- 거리 벡터 알고리즘 : 이웃 라우터와만 라우팅 정보를 교환한다.
  - `벨만-포드 알고리즘`을 기반으로 하여 분산 계산한다. 
> 자신에게 직접 연결된 **이웃으로의 링크 비용과 그 이웃들로부터 수신하는 정보를 이용**한다.

**Comment by choiyoung69** [comment]
## Q1
거리 벡터 알고리즘에서 링크 비용이 증가할 때 **라우팅 루프**가 발생할 수 있습니다. 이런 라우팅 루프는 왜 발생하는 것이며 라우팅 루프를 방지하기 위해서는 어떤 방법을 활용해야 할까요?

## Q2
BGP는 경로를 선택할 때 보통 뜨거운 감자 라우팅보다는 더 복잡한 알고리즘을 사용합니다. 이때 사용하는 제거 규칙을 간단하게 설명해주세요

**Reply by H-sooyeon** [reply]
> Q1 거리 벡터 알고리즘에서 링크 비용이 증가할 때 라우팅 루프가 발생할 수 있습니다. 이런 라우팅 루프는 왜 발생하는 것이며 라우팅 루프를 방지하기 위해서는 어떤 방법을 활용해야 할까요?

거리 벡터 알고리즘에서 각 라우터는 자신과 직졉 연결된 이웃 라우터들과 정보를 주고받는다.
그 정보를 바탕으로 자신의 라우팅 테이블을 업데이트하는데, 이때 정보 전달의 지연, 잘못된 경로 정보의 순환 등으로 인해 라우팅 루프가 발생하게 된다.
라우팅 루프를 방지하기 위해 Poison Reverse라는 방법을 사용할 수 있다.
라우터가 경로 정보를 받을 때 자신의 경로가 아니면 해당 경로를 '무한대'로 표시하여 다른 라우터가 이 경로를 사용하지 않도록 한다.
이를 통해 잘못된 경로가 확산되지 않도록 할 수 있다.

> Q2 BGP는 경로를 선택할 때 보통 뜨거운 감자 라우팅보다는 더 복잡한 알고리즘을 사용합니다. 이때 사용하는 제거 규칙을 간단하게 설명해주세요

적합하지 않은 경로들을 차례로 제거하다 최종적으로 남은 경로가 BGP에서 선택한 경로가 되는데,
우선순위가 높은 지역 선호도 값에 의해 차례로 제거된다.
즉, 최고 지역 선호 값을 가진 경로가 선택된다. 같은 값을 가진 경로가 여러 개 있다면 이들 중에서 최단 AS-PATH를 가진 경로가 선택된다.

**Reply by Starlight258** [reply]
> ## Q1
> 거리 벡터 알고리즘에서 링크 비용이 증가할 때 **라우팅 루프**가 발생할 수 있습니다. 이런 라우팅 루프는 왜 발생하는 것이며 라우팅 루프를 방지하기 위해서는 어떤 방법을 활용해야 할까요?

- 링크 비용이 증가해도 모든 라우터가 동시에 그 사실을 알지 못하므로 오래된 정보를 바탕으로 결정하는 라우터와 새로운 정보를 바탕으로 경로를 결정하는 라우터들 사이에서 라우팅 루프가 일어날 수 있다.
- 이를 방지하기 위해 `포이즌 리버스` (노드가 특정 목적지로 가는 경로로 사용하는 이웃 노드에게 해당 목적지까지의 거리를 무한대로 알리는 기법)을 사용할 수 있다.

> ## Q2
> BGP는 경로를 선택할 때 보통 뜨거운 감자 라우팅보다는 더 복잡한 알고리즘을 사용합니다. 이때 사용하는 제거 규칙을 간단하게 설명해주세요

주어진 목적지까지의 모든 경로 중에서 하나의 경로가 남을 때까지 제거 규칙을 수행한다.
- 1. 가장 높은 `지역 선호값`을 가진 경로를 선택한다.
- 2. `최단 AS-PATH`를 가진 경로가 선택된다.
- 3. 남은 경로들에 대해 `뜨거운 감자 라우팅(NEXT-HOP까지의 최단 경로)`을 수행한다.
	- `거리 벡터 알고리즘`을 사용하며 `AS 홉 수`를 거리값으로 사용한다.
- 4. `BGP 식별자`를 선택하여 경로를 선택한다.

**Comment by Starlight258** [comment]
# Q1
`BGP`와 `OSPF`를 간단히 설명해주세요.
두 프로토콜은 언제 사용되나요?

# Q2
`IP`의 `애니캐스트`에 대해 설명하고 어떤 시스템에서 사용되는지 설명해주세요.

**Reply by H-sooyeon** [reply]
> Q1 BGP와 OSPF를 간단히 설명해주세요. 두 프로토콜은 언제 사용되나요?

BGP는 자율 시스템(AS)간에 라우팅 정보를 교환한다. 확장성이 크고, 정책 기반 라우팅이 가능하다.
주로 인터넷 상에서 서로 다른 조직의 네트워크간에 연결을 설정할 때 사용된다.
ex) 여러 ISP들이 서로 네트워크 트래픽을 교환할 때 BGP를 사용해 최적의 경로를 선택하고 관리한다.

OSPF는 단일 조직 내의 네트워크 경로를 설정하기 위해 사용된다.
동일한 조직의 여러 라우터 간에 빠르고 효율적으로 경로를 설정 및 관리해야 하는 네트워크 내부에서 사용된다.
ex) 기업 내부 네트워크, 데이터센터 등에서 많이 사용된다.

<br />

> Q2 IP의 애니캐스트에 대해 설명하고 어떤 시스템에서 사용되는지 설명해주세요.

애니캐스트는 하나의 IP 주소가 여러 개의 서버에 할당되며 클라이언트가 해당 IP 주소로 요청을 보내면 가장 가까운 서버로 라우팅되는 방식이다.
클라이언트에게 최적의 경로로 서비스를 제공할 수 있도록 설계되어 있어, 지연시간을 줄이고 장애 발생 시 다른 애니캐스트 서버로 자동으로 요청이 분배되어 높은 가용성을 유지할 수 있다.

CDN은 애니캐스트를 통해 사용자와 가장 가까운 캐시 서버로 요청을 라우팅한다.
이를 통해 콘텐츠 전송 지연을 최소화하고, 네트워크 부하를 분산시킨다.

**Reply by Starlight258** [reply]
> # Q1
> `BGP`와 `OSPF`를 간단히 설명해주세요. 두 프로토콜은 언제 사용되나요?
> 

- `OSPF` : **단일 AS 내의 라우터들 간 라우팅 프로토콜**
  - `OSPF`를 이용하여 각 라우터는 전체 AS에 대한 토폴로지를 얻는다.
  - 단일 조직 내의 내부 네트워크 라우팅에 사용된다.
- `BGP` : **AS 간 라우팅 프로토콜**
  - 서브넷 주소 `prefix`로의 **가장 좋은 경로를 결정**한다.
  - 주로 서로 다른 조직이나 ISP 간의 대규모 외부 라우팅시에 사용한다.

> # Q2
> `IP`의 `애니캐스트`에 대해 설명하고 어떤 시스템에서 사용되는지 설명해주세요.

- 애니캐스트 : **동일한 IP 주소**를 여러 서버나 네트워크 장치에 할당하여 **네트워크 토폴로지 상 가장 가까운 위치로 라우팅**되도록 하는 방식이다.
- `DNS 시스템`에서는 DNS 질의를 `가장 가까운 루트 DNS 서버`로 전송하기 위해 사용한다.

**Comment by Hanjaemo** [comment]
### Q1
네트워크 트래픽을 제어하는 방식 중 라우터별 제어와 논리적 중앙 집중형 제어의 차이를 설명해 주세요.

### Q2
BGP가 AS 간 최적의 경로를 결정하는 방법에 대해 설명해 주세요.

**Reply by H-sooyeon** [reply]
> Q1 네트워크 트래픽을 제어하는 방식 중 라우터별 제어와 논리적 중앙 집중형 제어의 차이를 설명해 주세요.

라우터별 제어는 각 라우터가 가지고 있는 라우팅 테이블을 이웃 라우터들과 정보를 교환하며 갱신한다.
이를 기반으로 목적지까지의 최적 경로를 결정한다.
확장성이 뛰어나고 분산 네트워크 환경에서 유연하게 대처할 수 있다.

중앙 집중형 제어는 네트워크 제어를 중앙 집중식으로 관리하는 방식이다.
네트워크 내의 장비들은 단순히 패킷을 전달하는 역할만 하고, 중앙 컨트롤러가 모든 라우터 및 스위치의 트래픽 경로를 결정한다.
네트워크 상태를 중앙에서 모니터링하므로 유연한 트래픽 관리와 정교한 정책 적용이 가능하다.

<br />

> Q2 BGP가 AS 간 최적의 경로를 결정하는 방법에 대해 설명해 주세요.

BGP는 단순한 거리 기반 경로 선택보단 정책 기반으로 경로를 선택한다.
AS의 비즈니스 목적에 맞춰 트래픽 흐름을 제어할 수 있도록 하며 특정 AS를 통해 트래픽을 유도하거나 회피하는 정책을 설정할 수 있다.
또한, AS간의 트래픽 흐름이 항상 대칭적이지는 않다.
즉, 패킷이 출발지에서 목적지로 가는 경로와 목적지에서 출발지로 돌아오는 경로가 다를 수 있다.
이를 통해 BGP는 복잡한 네트워크 환경에서도 확장성 있고 유연한 경로 제어를 가능하게 한다.

**Reply by Starlight258** [reply]
> ### Q1
> 네트워크 트래픽을 제어하는 방식 중 라우터별 제어와 논리적 중앙 집중형 제어의 차이를 설명해 주세요.
> 
- **분산형 라우팅 알고리즘 (라우터별 제어)** : 각 노드는 자신에 직접 연결된 링크에 대한 비용 정보만 가진다.
	- `반복된 계산`과 `이웃 노드간의 정보 교환`을 통해 노드는 목적지까지의 **최소 비용 경로를 계산**한다.
- **중앙 집중형 라우팅 알고리즘** : 네트워크 전체에 대한 완전한 정보를 가지고 출발지와 목적지 사이의 최소 비용 경로를 계산한다.
	- 계산은 중앙 집중된 컨트롤러에서 수행하거나 각 라우팅 모듈로 복사될 수 있다.

> ### Q2
> BGP가 AS 간 최적의 경로를 결정하는 방법에 대해 설명해 주세요.
- 주어진 목적지까지의 모든 경로 중에서 하나의 경로가 남을 때까지 제거 규칙을 수행한다.
 1. 가장 높은 `지역 선호값`을 가진 경로를 선택한다.
 2. `최단 AS-PATH`를 가진 경로가 선택된다.
 3. 남은 경로들에 대해 `뜨거운 감자 라우팅(NEXT-HOP까지의 최단 경로)`을 수행한다.
	- `거리 벡터 알고리즘`을 사용하며 `AS 홉 수`를 거리값으로 사용한다.
 4. `BGP 식별자`를 선택하여 경로를 선택한다.
